ui/src/main/webapp/src/app/components/reports/application-details/application-details.component.ts

```diff
@@ -113,164 +96,95 @@ export class ApplicationDetailsComponent implements OnInit {
-                        });
-
-                        if (asArchive.discriminator.indexOf(IdentifiedArchiveModel.discriminator) != -1)
-                            this.traversalToSHA1.set(traversal.vertexId, asArchive.sHA1Hash);
-                    }
 
-                    if (this.duplicateProjects.has(canonical.vertexId)) {
-                        this.duplicateProjects.get(canonical.vertexId).push(traversal);
-                        return;
-                    } else {
-                        this.traversalToCanonical.set(traversal.vertexId, canonical);
-                        this.allProjects.push(traversal);
-                        this.duplicateProjects.set(canonical.vertexId, [traversal]);
-                    }
+            // Maintain a list by canonical id, so that we can display this on the canonical project
+            let traversalsForCanonicalID = this.traversalsForCanonicalVertexID.get(traversal.canonicalID);
+            if (!traversalsForCanonicalID)
+                traversalsForCanonicalID = [];
+            traversalsForCanonicalID.push(traversal);
+            traversalsForCanonicalID = traversalsForCanonicalID.sort(compareTraversals);
+            this.traversalsForCanonicalVertexID.set(traversal.canonicalID, traversalsForCanonicalID);
+
+            // If these do not match, then this is not a canonical project.
+            // We don't need to calculate points for this.
+            if (traversal.currentID != traversal.canonicalID)
+                return;
 
-                    this.storeProjectData(traversal);
-                });
-            });
+            this.allProjects.push(traversal);
+            this.storeProjectData(traversal);
+            this.storePointsForTraversal(traversal);
         });
     }
 
-    private getDuplicateProjectPaths(traversal:PersistedProjectModelTraversalModel):string[] {
-        let canonical = this.traversalToCanonical.get(traversal.vertexId);
-        if (!canonical)
-            return [];
-
-        let duplicateList = this.duplicateProjects.get(canonical.vertexId);
-        if (!duplicateList)
-            return [];
-
-        let names = duplicateList.map(traversal => traversal.path);
-        return names.sort(comparePaths);
+    private hasDuplicateProjects(traversal:ProjectTraversalReducedDTO):boolean {
+        return this.traversalsForCanonicalVertexID.get(traversal.canonicalID).length > 1;
     }
 
-    private isDuplicateProject(traversal:PersistedProjectModelTraversalModel):boolean {
-        let canonical = this.traversalToCanonical.get(traversal.vertexId);
-        if (!canonical)
-            return false;
-
-        let duplicateList = this.duplicateProjects.get(canonical.vertexId);
-        if (!duplicateList)
-            return false;
-
-        return duplicateList.length > 1;
+    resolveString(id:number):string {
+        if (id == 0)
+            return null;
+        return this.applicationDetails.stringCache.byID[id];
```

Comment:
This is optimization of data format transferred from backend. I think service should take care of this and it should be transparent for component.